service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,318 @@
+	s.ctrl = host.NewController(ServiceName)
+
+	s.fetch()
+
+	// Run the service's own controller.
+	go func() {
+		s.ctrl.Run(ctx)
+	}()
+
+	for {
+		select {
+		case <-s.ch:
+			b, err := s.getAPIConfig()
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to fetch configuration from the API", "err", err)
+				continue
+			}
+
+			// The polling loop got the same configuration, no need to reload.
+			newConfigHash := getHash(b)
+			if s.getCfgHash() == newConfigHash {
+				level.Debug(s.opts.Logger).Log("msg", "skipping to the next polling loop")
+				continue
+			}
+
+			// We have a new configuration let's try to load it
+			err = s.parseAndLoad(b)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to load configuration from the API", "err", err)
+				continue
+			}
+
+			level.Info(s.opts.Logger).Log("msg", "new remote configuration loaded successfully")
+
+			// If successful, flush to disk and keep a copy.
+			err = os.WriteFile(s.dataPath, b, 0750)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to flush remote configuration contents the on-disk cache", "err", err)
+			}
+			s.setCfgHash(newConfigHash)
+		case <-ctx.Done():
+			s.ticker.Stop()
+			return nil
+		}
+	}
+}
+
+// Update implements [service.Service] and applies settings.
+func (s *Service) Update(newConfig any) error {
+	newArgs := newConfig.(Arguments)
+
+	// We either never set the block on the first place, or recently removed
+	// it. Make sure we stop everything gracefully before returning.
+	if newArgs.URL == "" {
+		s.mut.Lock()
+		s.ch = nil
+		s.ticker.Reset(math.MaxInt64)
+		s.asClient = noopClient{}
+		s.args.HTTPClientConfig = config.CloneDefaultHTTPClientConfig()
+		s.mut.Unlock()
+		return nil
+	}
+
+	s.mut.Lock()
+	hash, err := newArgs.Hash()
+	if err != nil {
+		return err
+	}
+	s.dataPath = filepath.Join(s.opts.StoragePath, ServiceName, hash)
+	s.ticker.Reset(newArgs.PollFrequency)
+	s.ch = s.ticker.C
+	// Update the HTTP client last since it might fail.
+	if !reflect.DeepEqual(s.args.HTTPClientConfig, newArgs.HTTPClientConfig) {
+		httpClient, err := commonconfig.NewClientFromConfig(*newArgs.HTTPClientConfig.Convert(), "remoteconfig")
+		if err != nil {
+			return err
+		}
+		s.asClient = agentv1connect.NewAgentServiceClient(
+			httpClient,
+			newArgs.URL,
+		)
+	}
+	s.args = newArgs // Update the args as the last step to avoid polluting any comparisons
+	s.mut.Unlock()
+
+	// If we've already called Run, then immediately trigger an API call with
+	// the updated Arguments, and/or fall back to the updated cache location.
+	if s.ctrl != nil && s.ctrl.Ready() {
+		s.fetch()
+	}
+
+	return nil
+}
+
+// fetch attempts to read configuration from the API and the local cache
+// and then parse/load in order of preference.
+func (s *Service) fetch() {
+	if !s.isEnabled() {
+		return
+	}
+
+	var b1, b2 []byte
+	b1, err := s.getAPIConfig()
```

Comment:
I wonder if the reason the multiple errors are needed is due to the algorithm order. I think what you have is working, but this may clear it up:

1) getAPIConfig
2) if 1 succeeds, attempt parseAndLoad the result from 1
3) if 1 or 2 fails, getCachedConfig
4) if 3 succeeds, parseAndLoad the result from 3

We could optimize that getAPIConfig and getCachedConfig are called in parallel with go routines if we want.

Reply:
@wildum wrote a version of this below*

https://github.com/grafana/agent/pull/6277/files#r1481818052

Reply:
I'm ok with either way, but I don't think it makes sense to introduce more complex optimizations at this stage.